File: UnitB/ch03/data3.3-atlantcommute.jl

## 1. load package
include("../../utils.jl")
  using GLMakie,CSV,DataFrames,Random,StatsBase,PrettyTables,Pipe

##  2. load data
  #= 
    feature :["City", "Age", "Distance", "Time", "Sex"]
  =#
  desc=Lock5Table(270,"CommuteAtlanta","data3.1",["City", "Age", "Distance", "Time", "Sex"])
  df=@pipe load_csv(desc.name)|>select(_,desc.feature[4])
#= 
┌─────────┬───────┬──────────┬───────┬─────────┐
│    City │   Age │ Distance │  Time │     Sex │
│ String7 │ Int64 │    Int64 │ Int64 │ String1 │
├─────────┼───────┼──────────┼───────┼─────────┤
│ Atlanta │    19 │       10 │    15 │       M │
│ Atlanta │    55 │       45 │    60 │       M │
│ Atlanta │    48 │       12 │    45 │       M │
│ Atlanta │    45 │        4 │    10 │       F │
│ Atlanta │    48 │       15 │    30 │       F │
└─────────┴───────┴──────────┴───────┴─────────┘
=#

function plot_dotplot(gdf::GroupedDataFrame;title="CommuteAtlanta Time sampling")
  xs=@pipe keys(gdf).|>values(_)[1]
  ys=[nrow(g) for g in gdf]
  fig=Figure(resolution=(1200,500))
  ax=Axis(fig[1,1],xlabel="Time",ylabel="Count",title=title)
  Box(fig[1,1];color =(:orange,0.1),strokewidth=0.2)

  for (idx, x) in enumerate(xs)
      
      for y in 1:ys[idx]
          
          scatter!(ax,x,y
          ;marker=:circle,markersize=10,color=:black
          ,strokewidth=0.1,strokecolor=:black)
      end
  end
  
 fig
end


fig=groupby(df,:Time)|>plot_dotplot
#save("data3.3-atlantcommute-dotplot.png",fig)

File: marimo/ch03/data3.3-atlantcommute.py
import marimo

__generated_with = "0.24.0"
app = marimo.App(width="medium")


@app.cell
def _():
    import marimo as mo
    import sys
    from pathlib import Path
    import numpy as np
    import polars as pl
    import matplotlib.pyplot as plt
    from scipy import stats
    from great_tables import GT
    _p = Path(mo.notebook_location()).resolve().parent
    while _p != _p.parent and not (_p / "lock5.py").exists():
        _p = _p.parent
    if str(_p) not in sys.path:
        sys.path.insert(0, str(_p))
    import lock5
    return mo, sys, Path, np, pl, plt, stats, GT, lock5


@app.cell
def _(mo):
    mo.md(r"""# Data 3.3 — CommuteAtlanta time dot plot

Port of `UnitB/ch03/data3.3-atlantcommute.jl`. Dot plot of the `Time` variable.""")
    return


@app.cell
def _(lock5):
    desc = lock5.Lock5Table(270, "CommuteAtlanta", "data3.1",
                            ["City", "Age", "Distance", "Time", "Sex"])
    df = lock5.load_csv(desc.name).select(desc.feature[3])
    times = df["Time"].to_numpy().astype(float)
    return desc, df, times


@app.cell
def _(lock5, times):
    fig, ax = lock5.plot_dotplot(times, nbins=len(set(times)) if len(set(times)) < 60 else 30,
                                 title="CommuteAtlanta Time sampling",
                                 xlabel="Time", ylabel="Count", color="black")
    fig
    return fig, ax


if __name__ == "__main__":
    app.run()
